File: examples/material/js/material-example.js
const defaultMaterial = WebSG.getMaterialByName("Default");
const groundMaterial = WebSG.getMaterialByName("Ground");
const glassMaterial = WebSG.getMaterialByName("Glass");
const groundTexture = groundMaterial.baseColorTexture;

let elapsed = 0;

onupdate = (dt) => {
  elapsed += dt;

  defaultMaterial.baseColorFactor[1] = (Math.sin(elapsed) + 1) / 2;
  groundMaterial.baseColorFactor[0] = (Math.sin(elapsed) + 1) / 2;
  glassMaterial.baseColorFactor[3] = (Math.sin(elapsed) + 1) / 2;

  if (elapsed > 5 && groundMaterial.baseColorTexture) {
    console.log("test", groundTexture);
    defaultMaterial.baseColorTexture = groundTexture;
    groundMaterial.baseColorTexture = null;
  }
};

File: examples/material/c/src/material-example.c
#include <math.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#include "websg.h"

export void *websg_allocate(int size) {
  return malloc(size);
}

export void websg_deallocate(void *ptr) {
  free(ptr);
}

Material *default_material;
Material *ground_material;
Material *glass_material;
Texture *ground_texture;
float_t elapsed = 0;

export void websg_loaded() {
  default_material = websg_get_resource_by_name(ResourceType_Material, "Default");
  ground_material = websg_get_resource_by_name(ResourceType_Material, "Ground");
  glass_material = websg_get_resource_by_name(ResourceType_Material, "Glass");
  ground_texture = ground_material->base_color_texture;
}

export void websg_update(float_t dt) {
  elapsed += dt;

  default_material->base_color_factor[1] = (sin(elapsed) + 1) / 2;
  ground_material->base_color_factor[0] = (sin(elapsed) + 1) / 2;
  glass_material->base_color_factor[3] = (sin(elapsed) + 1) / 2;

  if (elapsed > 5 && ground_material->base_color_texture) {
    default_material->base_color_texture = ground_texture;
    ground_material->base_color_texture = 0;
  }
}